File: graylog2-web-interface/src/util/StringUtils.test.ts
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */

import StringUtils from 'util/StringUtils';

describe('capitalizeFirstLetter', () => {
  it('capitalizes the first letter of a string supplied', () => {
    const testString = 'capitalizeMe';
    const expectedOutcome = 'CapitalizeMe';

    const newString = StringUtils.capitalizeFirstLetter(testString);

    expect(newString).toBe(expectedOutcome);
  });
});

describe('HTML escape unescape', () => {
  const unescapedString = 'Hello<enter-name> & enjoy';
  const escapedString = 'Hello&lt;enter-name&gt; &amp; enjoy';

  it('escapes a string for HTML use', () => {
    const result = StringUtils.escapeHTML(unescapedString);

    expect(result).toBe(escapedString);
  });

  it('unescapes a string used in HTML', () => {
    const result = StringUtils.unescapeHTML(escapedString);

    expect(result).toBe(unescapedString);
  });
});

describe('pluralize', () => {
  it('pluralizes a string', () => {
    const singular = 'racecar';
    const plural = 'racecars';
    const resultSingular = StringUtils.pluralize(1, singular, plural);

    expect(resultSingular).toBe(singular);

    const resultPlural = StringUtils.pluralize(2, singular, plural);

    expect(resultPlural).toBe(plural);
  });
});

describe('stringify', () => {
  const testObj = {
    fruitName: 'Apple',
    brixRating: 7,
    color: 'red',
  };
  const stringifiedObj = '{"fruitName":"Apple","brixRating":7,"color":"red"}';

  it('stringifies an object', () => {
    const result = StringUtils.stringify(testObj);

    expect(result).toBe(stringifiedObj);
  });

  it('returns anything else as a string', () => {
    const testNum = 7;
    const testArray = [1, 2, 3];

    const numResult = StringUtils.stringify(testNum);

    expect(numResult).toBe('7');

    const arrayResult = StringUtils.stringify(testArray);

    expect(arrayResult).toBe('[1,2,3]');
  });

  it('returns an empty string if String() would return an error', () => {
    const canNotStringMe = undefined;
    const result = StringUtils.stringify(canNotStringMe);

    expect(result).toBe('undefined');
  });
});

describe('replaceSpaces', () => {
  it('replaces spaces in a string with "-"', () => {
    const makeHyphenized = 'Rubber baby bunky bumpers';
    const hyphenized = 'Rubber-baby-bunky-bumpers';
    const result = StringUtils.replaceSpaces(makeHyphenized);

    expect(result).toBe(hyphenized);
  });

  it('replaces spaces in a string with a supplied character', () => {
    const makePipified = 'Rubber baby bunky bumpers';
    const pipified = 'Rubber|baby|bunky|bumpers';
    const result = StringUtils.replaceSpaces(makePipified, '|');

    expect(result).toBe(pipified);
  });
});

describe('toTitleCase', () => {
  it('title cases a string and defaults the split to " "', () => {
    const makeTitleCase = 'Rubber baby bunky bumpers.';
    const titleCased = 'Rubber Baby Bunky Bumpers.';
    const result = StringUtils.toTitleCase(makeTitleCase);

    expect(result).toBe(titleCased);
  });

  it('titles cases a string based on supplied split character', () => {
    const makeTitleCase = 'Rubber-baby-bunky-bumpers.';
    const titleCased = 'Rubber Baby Bunky Bumpers.';
    const result = StringUtils.toTitleCase(makeTitleCase, '-');

    expect(result).toBe(titleCased);
  });
});

describe('truncateWithEllipses', () => {
  it('truncates a string with a default length of 10 and ending string of "..."', () => {
    const makeShorter = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent nec.';
    const truncated = 'Lorem ipsu...';
    const result = StringUtils.truncateWithEllipses(makeShorter);

    expect(result).toBe(truncated);
  });

  it('truncates a string with supplied length and default ending string of "..."', () => {
    const makeShorter = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent nec.';
    const truncated = 'Lorem ipsum dol...';
    const result = StringUtils.truncateWithEllipses(makeShorter, 15);

    expect(result).toBe(truncated);
  });

  it('truncates a string with supplied length and supplied ending string', () => {
    const makeShorter = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent nec.';
    const truncated = 'Lorem ipsum dol|||';
    const result = StringUtils.truncateWithEllipses(makeShorter, 15, '|||');

    expect(result).toBe(truncated);
  });
});

File: graylog2-server/src/test/java/org/graylog2/shared/utilities/StringUtilsTest.java
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package org.graylog2.shared.utilities;

import org.assertj.core.api.Assertions;
import org.junit.Test;

import java.util.Collections;
import java.util.List;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.graylog2.shared.utilities.StringUtils.toLowerCase;
import static org.graylog2.shared.utilities.StringUtils.toUpperCase;

public class StringUtilsTest {
    @Test
    public void testToUpperCase() {
        //noinspection DataFlowIssue
        assertThatThrownBy(() -> toUpperCase(null)).isInstanceOf(IllegalArgumentException.class);
        assertThat(toUpperCase("")).isEqualTo("");
        assertThat(toUpperCase(" ")).isEqualTo(" ");
        assertThat(toUpperCase("Hello")).isEqualTo("HELLO");
    }

    @Test
    public void testToLowerCase() {
        //noinspection DataFlowIssue
        assertThatThrownBy(() -> toLowerCase(null)).isInstanceOf(IllegalArgumentException.class);
        assertThat(toLowerCase("")).isEqualTo("");
        assertThat(toLowerCase(" ")).isEqualTo(" ");
        assertThat(toLowerCase("Hello")).isEqualTo("hello");
    }

    @Test
    public void testHumanReadable() {
        assertThat(StringUtils.humanReadableByteCount(1024L * 1024L * 1024L * 5L + 1024L * 1024L * 512L)).isEqualTo("5.5 GiB");
        assertThat(StringUtils.humanReadableByteCount(1024L * 1024L * 1024L * 5L)).isEqualTo("5.0 GiB");
        assertThat(StringUtils.humanReadableByteCount(1024L * 1024L * 4L + 1024L * 900L)).isEqualTo("4.9 MiB");
        assertThat(StringUtils.humanReadableByteCount(1023)).isEqualTo("1023 B");
        assertThat(StringUtils.humanReadableByteCount(1024)).isEqualTo("1.0 KiB");
        assertThat(StringUtils.humanReadableByteCount(1024L * 1024L * 1024L * 1024L * 5L + 1024L * 1024L * 512L)).isEqualTo("5.0 TiB");
        assertThat(StringUtils.humanReadableByteCount(1024L * 5L + 512L)).isEqualTo("5.5 KiB");
    }

    @Test
    public void testSplitByComma() {
        Assertions.assertThat(StringUtils.splitByComma((Set<String>) null))
                .isInstanceOf(Set.class)
                .isEmpty();

        Assertions.assertThat(StringUtils.splitByComma(Collections.emptySet()))
                .isInstanceOf(Set.class)
                .isEmpty();

        Assertions.assertThat(StringUtils.splitByComma(List.of("one", "two,three,", "", " ")))
                .hasSize(3)
                .containsExactlyInAnyOrder("one", "two", "three");


        Assertions.assertThat(StringUtils.splitByComma(List.of("one", "two,three")))
                .hasSize(3)
                .containsExactlyInAnyOrder("one", "two", "three");


        Assertions.assertThat(StringUtils.splitByComma(List.of("one", "two,three")))
                .hasSize(3)
                .containsExactlyInAnyOrder("one", "two", "three");


    }
}
